README
About
-----

Python Module for manipulating SMPTE timecode. Supports 23.98, 24, 25, 29.97,
30, 50, 59.94, 60 frame rates and milliseconds (1000 fps).

This library is a fork of the original PyTimeCode python library. You should
not use the two library together (PyTimeCode is not maintained and has known
bugs).

The math behind the drop frame calculation is based on the
`blog post of David Heidelberger`_ .

.. _`blog post of David Heidelberger`: http://www.davidheidelberger.com/blog/?p=29

Simple math operations like, addition, subtraction, multiplication or division
with an integer value or with a timecode is possible. Math operations between
timecodes with different frame rates are supported. So:

    from timecode import Timecode
  
    tc1 = Timecode('29.97', '00:00:00:01')
    tc2 = Timecode('24', '00:00:00:10')
    tc3 = tc1 + tc2
    assert tc3.framerate == '29.97'
    assert tc3.frames == 11
    assert tc3 == '00:00:00:11'

Creating a Timecode instance with a start timecode of '00:00:00:00' will
result a timecode object where the total number of frames is 0. So::

    tc4 = Timecode('24', '00:00:00:00')
    assert tc4.frames == 0

Use the ``frame_number`` attribute if you want to get a 1 based frame number::

    assert tc4.frame_number == 1

Frame rates 29.97 and 59.94 are always drop frame, and all the others are non
drop frame.

The SMPTE standard limits the timecode to 24 hours. All maths performed by
Timecode is therefore performed modulo ``one_day``.

Please report any bugs to the `GitHub`_ page.

.. _`GitHub`: https://github.com/Nick-Shaw/timecode

Copyright 2014 Joshua Banton and PyTimeCode developers.


CHANGELOG
=======
Changes
=======

Antler 0.2.0
============

* **Update:** ``self.frame`` and ``self.frame_number`` have been reversed so
that maths matches other timcode implementations. ``00:00:00:10 + 00:00:00:10
== 00:00:00:20`` rather than ``00:00:00:21`` as previously.

* **New:** Incomplete timecode strings, such as ``10:00`` can be entered.

* **New:** Frame counts can be entered, so ``Timecode(25, 100)`` becomes
``00:00:04:00``.

* **Update:** All maths is now performed modulo ``self.one_day``, so it wraps
around at 24 hours.

0.4.0
=====

* **New:** Frame delimiter is now set to ":" for Non Drop Frame, ";" for Drop
  Frame and "." for millisecond based time codes.
  If ``Timecode.__init__()`` start_timecode is passed a string with the wrong
  delimiter it will be converted automatically.

* **Update:** All tests involving Drop Frame and millisecond time codes are now
  set to use the new delimiter.

* **New:** ``Timecode.tc_to_string()`` method added to present the correctly
  formatted time code.

* **New:** ``Timecode.ms_frame`` boolean attribute added.

* **New:** ``Timecode.__init__()`` now supports strings, ints and floats for
  the framerate argument.

0.3.0
=====

* **New:** Renamed the library to ``timecode``.

0.2.0
=====

* **New:** Rewritten the whole library from scratch.

* **New:** Most important change is the licencing. There was now license
  defined in the previous implementation. The library is now licensed under MIT
  license.

* **Update:** Timecode.__init__() arguments has been changed, removed the
  unnecessary ``drop_frame``, ``iter_returns`` arguments.

  Drop frame can be interpreted from the ``framerate`` argument and
  ``iter_returns`` is unnecessary cause any iteration on the object will return
  another ``Timecode`` instance.

  If you want to get a string representation use ``Timecode.__str__()`` or
  ``str(Timecode)`` or ``Timecode.__repr__()`` or ``\`Timecode\``` or
  ``'%s' % Timecode`` any other thing that will convert it to a string.

  If you want to get an integer use ``Timecode.frames`` or
  ``Timecode.frame_count`` depending on what you want to get out of it.

  So setting the ``iter_returns`` to something and nailing the output was
  unnecessary.

* **Update:** Updated the drop frame calculation to a much better one, which
  is based on to the blog post of David Heidelberger at
  http://www.davidheidelberger.com/blog/?p=29

* **New:** Added ``Timecode.__eq__()`` so it is now possible to check the
  equality of two timecode instances or a timecode and a string or a timecode
  and an integer (which will check the total frame count).

* **Update:** ``Timecode.tc_to_frames()`` now needs a timecode as a string
  and will return an integer value which is the number of frames in that
  timecode.

* **Update:** ``Timecode.frames_to_tc()`` now needs an integer frame count
  and returns 4 integers for hours, minutes, seconds and frames.

* **Update:** ``Timecode.hrs``, ``Timecode.mins``, ``Timecode.secs`` and
  ``Timecode.frs`` attributes are now properties. Because it was so rare to
  check the individual hours, minutes, seconds or frame values, their values
  are calculated with ``Timecode.frames_to_tc()`` method. But in future they
  can still be converted to attributes and their value will be updated each
  time the ``Timecode.frames`` attribute is changed (so add a ``_frames``
  attribute and make ``frames`` a property with a getter and setter, and update
  the hrs, mins, secs and frs in setter etc.).

* **Update:** Removed ``Timecode.calc_drop_frame()`` method. The drop frame
  calculation is neatly done inside ``Timecode.frames_to_tc()`` and
  ``Timecode.tc_to_frames()`` methods.

* **Update:** Updated ``Timecode.parse_timecode()`` method to a much simpler
  algorithm.

* **Update:** Removed ``Timecode.__return_item__()`` method. It is not
  necessary to return an item in that way anymore.

* **Update:** Removed ``Timecode.make_timecode()`` method. It was another
  unnecessary method, so it is removed. Now using simple python string
  templates for string representations.

* **New:** Added ``timecode.__version__`` string, and set the value to
  "0.2.0".

* **Update:** Removed ``Timecode.set_int_framerate()`` method. Setting the
  framerate will automatically set the ``Timecode.int_framerate`` attribute.


timecode/__init__.py
#!-*- coding: utf-8 -*-
# The MIT License (MIT)
#
# Original Copyright (c) 2014 Joshua Banton and PyTimeCode developers
# This version Copyright (c) 2017 Nick Shaw, Antler Post
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.


__version__ = 'Antler 0.2.0'


class Timecode(object):
    def __init__(self, framerate, start_timecode=None, start_seconds=None,
                 frames=None):
        """The main timecode class.

        Does all the calculation over frames, so the main data it holds is
        frames, then when required it converts the frames to a timecode by
        using the frame rate setting.

        :param framerate: The frame rate of the Timecode instance. It
          should be one of ['23.976', '23.98', '24', '25', '29.97', '30', '50',
          '59.94', '60', 'ms'] where "ms" equals to 1000 fps. Can not be
          skipped. Setting the framerate will automatically set the
          :attr:`.drop_frame` attribute to correct value.
        :param start_timecode: The start timecode. Use this to be able to
          set the timecode of this Timecode instance. It can be skipped and
          then the frames attribute will define the timecode, and if it is also
          skipped then the start_second attribute will define the start
          timecode, and if start_seconds is also skipped then the default value
          of '00:00:00:00' will be used.
        :type framerate: str or int or float
        :type start_timecode: str or None
        :param start_seconds: A float or integer value showing the seconds.
        :param int frames: Timecode objects can be initialized with an
          integer number showing the total frames.
        """
        self.drop_frame = False
        self.ms_frame = False
        self._int_framerate = None
        self._framerate = None
        self.framerate = framerate

        self.frames = None

        # attribute override order
        # start_timecode > frames > start_seconds
        if start_timecode:
            if isinstance(start_timecode, str):
                self.frames = self.tc_to_frames(start_timecode)
            elif isinstance(start_timecode, int):
                self.frames = start_timecode
            elif isinstance(start_timecode, float):
                self.frames = self.float_to_tc(start_timecode)
            else:
                self.frames = 0
        else:
            if frames is not None:  # because 0==False, and frames can be 0
                self.frames = frames
            elif start_seconds is not None:
                self.frames = self.float_to_tc(start_seconds)
            else:
                # use default value of 00:00:00:00
                self.frames = self.tc_to_frames('00:00:00:00')

    @property
    def framerate(self):
        """getter for _framerate attribute
        """
        return self._framerate

    @framerate.setter
    def framerate(self, framerate):  # lint:ok
        """setter for the framerate attribute
        :param framerate:
        :return:
        """

        # check if number is passed and if so convert it to a string
        if isinstance(framerate, (int, float)):
            framerate = str(framerate)

        # set the int_frame_rate
        if framerate == '29.97':
            self._int_framerate = 30
            self.drop_frame = True
        elif framerate == '59.94':
            self._int_framerate = 60
            self.drop_frame = True
        elif framerate == '23.98' or framerate == '23.976':
            self._int_framerate = 24
        elif framerate in ['ms', '1000']:
            self._int_framerate = 1000
            self.ms_frame = True
            framerate = 1000
        elif framerate == 'frames':
            self._int_framerate = 1
        else:
            self._int_framerate = int(framerate)

        self._framerate = framerate

    def set_timecode(self, timecode):
        """Sets the frames by using the given timecode
        """
        self.frames = self.tc_to_frames(timecode)

    def float_to_tc(self, seconds):
        """set the frames by using the given seconds
        """
        return int(seconds * self._int_framerate)

    def tc_to_frames(self, timecode):
        """Converts the given timecode to frames
        """
        hours, minutes, seconds, frames = map(int,
                                              self.parse_timecode(timecode))

        ffps = float(self._framerate)

        if self.drop_frame:
            # Number of drop frames is 6% of framerate rounded to nearest
            # integer
            drop_frames = int(round(ffps * .066666))
        else:
            drop_frames = 0

        # We don't need the exact framerate anymore, we just need it rounded to
        # nearest integer
        ifps = self._int_framerate

        # Number of frames per hour (non-drop)
        hour_frames = ifps * 60 * 60

        # Number of frames per minute (non-drop)
        minute_frames = ifps * 60

        # Total number of minutes
        total_minutes = (60 * hours) + minutes

        frame_number = \
            ((hour_frames * hours) + (minute_frames * minutes) +
             (ifps * seconds) + frames) - \
            (drop_frames * (total_minutes - (total_minutes // 10)))

        frames = frame_number  # + 1

        return frames

    def frames_to_tc(self, frames):
        """Converts frames back to timecode

        :returns str: the string representation of the current time code
        """
        ffps = float(self._framerate)

        if self.drop_frame:
            # Number of frames to drop on the minute marks is the nearest
            # integer to 6% of the framerate
            drop_frames = int(round(ffps * .066666))
        else:
            drop_frames = 0

        # Number of frames in an hour
        frames_per_hour = int(round(ffps * 60 * 60))
        # Number of frames in a day - timecode rolls over after 24 hours
        frames_per_24_hours = frames_per_hour * 24
        # Number of frames per ten minutes
        frames_per_10_minutes = int(round(ffps * 60 * 10))
        # Number of frames per minute is the round of the framerate * 60 minus
        # the number of dropped frames
        frames_per_minute = int(round(ffps) * 60) - drop_frames

        frame_number = frames  # - 1

        if frame_number < 0:
            # Negative time. Add 24 hours.
            frame_number += frames_per_24_hours

        # If frame_number is greater than 24 hrs, next operation will rollover
        # clock
        frame_number %= frames_per_24_hours

        if self.drop_frame:
            d = frame_number // frames_per_10_minutes
            m = frame_number % frames_per_10_minutes
            if m > drop_frames:
                frame_number += (drop_frames * 9 * d) + \
                    drop_frames * ((m - drop_frames) // frames_per_minute)
            else:
                frame_number += drop_frames * 9 * d

        ifps = self._int_framerate
        frs = frame_number % ifps
        secs = (frame_number // ifps) % 60
        mins = ((frame_number // ifps) // 60) % 60
        hrs = (((frame_number // ifps) // 60) // 60)

        return hrs, mins, secs, frs

    def tc_to_string(self, hrs, mins, secs, frs):
        if self.frame_delimiter == '.':
            return "%02d:%02d:%02d%s%03d" % (hrs,
                                             mins,
                                             secs,
                                             '.',
                                             frs)
        else:
            return "%02d:%02d:%02d%s%02d" % (hrs,
                                             mins,
                                             secs,
                                             self.frame_delimiter,
                                             frs)

    @classmethod
    def parse_timecode(cls, timecode):
        """parses timecode string NDF '00:00:00:00' or DF '00:00:00;00' or
        milliseconds '00:00:00.000'
        """
        bfr = timecode.replace(';', ':').replace('.', ':').split(':')
        bfr.reverse()
        while len(bfr) < 4:
            # fill missing leading fields
            bfr.append('00')
        bfr.reverse()
        if '.' in timecode:
            # round/pad to three decimal places
            bfr[3] = str(int(round(float(bfr[3])/10**(len(bfr[3])-3))))
        hrs = int(bfr[0])
        mins = int(bfr[1])
        secs = int(bfr[2])
        frs = int(bfr[3])
        return hrs, mins, secs, frs

    @property
    def frame_delimiter(self):
        """Return correct symbol based on framerate."""
        if self.drop_frame:
            return ';'

        elif self.ms_frame:
            return '.'

        else:
            return ':'

    def __iter__(self):
        return self

    def next(self):
        self.add_frames(1)
        return self

    def back(self):
        self.sub_frames(1)
        return self

    def add_frames(self, frames):
        """adds or subtracts frames number of frames
        """
        self.frames += frames
        self.frames = self.frames % self.one_day

    def sub_frames(self, frames):
        """adds or subtracts frames number of frames
        """
        self.add_frames(-frames)

    def mult_frames(self, frames):
        """multiply frames
        """
        self.frames *= frames
        self.frames = self.frames % self.one_day

    def div_frames(self, frames):
        """adds or subtracts frames number of frames"""
        self.frames = self.frames / frames
        self.frames = self.frames % self.one_day

    def __eq__(self, other):
        """the overridden equality operator
        """
        if isinstance(other, Timecode):
            return self._framerate == other._framerate and \
                self.frames == other.frames
        elif isinstance(other, str):
            new_tc = Timecode(self._framerate, other)
            return self.__eq__(new_tc)
        elif isinstance(other, int):
            return self.frames == other

    def __add__(self, other):
        """returns new Timecode instance with the given timecode or frames
        added to this one
        """
        # duplicate current one
        tc = Timecode(self._framerate, frames=self.frames)

        if isinstance(other, Timecode):
            tc.add_frames(other.frames)
        elif isinstance(other, int):
            tc.add_frames(other)
        else:
            raise TimecodeError(
                'Type %s not supported for arithmetic.' %
                other.__class__.__name__
            )

        tc.frames = tc.frames % self.one_day
        return tc

    def __sub__(self, other):
        """returns new Timecode object with added timecodes"""
        if isinstance(other, Timecode):
            subtracted_frames = self.frames - other.frames
        elif isinstance(other, int):
            subtracted_frames = self.frames - other
        else:
            raise TimecodeError(
                'Type %s not supported for arithmetic.' %
                other.__class__.__name__
            )

        return Timecode(self._framerate, frames=(subtracted_frames % self.one_day))

    def __mul__(self, other):
        """returns new Timecode object with added timecodes"""
        if isinstance(other, Timecode):
            multiplied_frames = self.frames * other.frames
        elif isinstance(other, int):
            multiplied_frames = self.frames * other
        else:
            raise TimecodeError(
                'Type %s not supported for arithmetic.' %
                other.__class__.__name__
            )

        return Timecode(self._framerate, frames=(multiplied_frames % self.one_day))

    def __div__(self, other):
        """returns new Timecode object with added timecodes"""
        if isinstance(other, Timecode):
            div_frames = self.frames / other.frames
        elif isinstance(other, int):
            div_frames = self.frames / other
        else:
            raise TimecodeError(
                'Type %s not supported for arithmetic.' %
                other.__class__.__name__
            )

        return Timecode(self._framerate, frames=(div_frames % self.one_day))

    def __repr__(self):
        return self.tc_to_string(*self.frames_to_tc(self.frames))

    @property
    def hrs(self):
        hrs, mins, secs, frs = self.frames_to_tc(self.frames)
        return hrs

    @property
    def mins(self):
        hrs, mins, secs, frs = self.frames_to_tc(self.frames)
        return mins

    @property
    def secs(self):
        hrs, mins, secs, frs = self.frames_to_tc(self.frames)
        return secs

    @property
    def frs(self):
        hrs, mins, secs, frs = self.frames_to_tc(self.frames)
        return frs

    @property
    def frame_number(self):
        """returns the 1 based frame number of the current timecode instance
        """
        return self.frames + 1

    @property
    def one_day(self):
        """returns the number of frames in a day
        """
        return Timecode(self.framerate, '24:00:00:00').frames


class TimecodeError(Exception):
    """Raised when an error occurred in timecode calculation
    """
    pass
